content/blog/2023-12-11-circ-bitwidth-adjustment

```diff
@@ -0,0 +1,115 @@
+
+The current system represents basic elements as bitvectors, and is designed to adjust their bitwidth after each arithmetic operation. Based on the classification of the elements above, our projects reduces the number of constraints in the system by greedily deciding when to adjust bitwidth for elements falls under the second category while ensuring no overflows can happen in later operations involving the elements. Or in other words, adjustments are only introduced if the next operation is going to result in an overflow.
+
+Take this short program as an example:
+```
+def main(u32 a, u32 b) -> u32:
+    u32 c = a + b
+    u32 d = a + c
+    return c * d
+```
+
+The old compiler would adjust immediately at `a+b`, `a+c` and `c * d`. By applying our optimization, the system would treat c and d as intermediate elements and only adjust at c*d, since it is producing the final program output. 
+Without our optimization, the cost spent on adjusting the bitwidth will be: 33 constraint to adjust `a + b` and `a + c` respectively, and 64 to adjust `c * d`, which result in 130 constraints in total.
+In contrast, after our optimization, the cost spent on adjusting `a + b` and `a + c` would be 0, and 66 to adjust `c * d`, which result in 66 constraints in total. 
+
+## Implementation
+1. We create a new type of operations: `add_unadjust` and `mul_unadjust` for addition and multiplication to notify the constraint generation system not to adjust the bitwidth.
+2. We implemented two passes to rewrite the operations. First pass is to classify elements in the system. We implemented an initial pass to scan the entire program and store elements requiring adjustments in a set S. Whenever we encounter an operation that definitely requires adjustment, we add the elements used in the operation to S. Then we implemented a rewrite pass to rewrite operations done on bitvectors not included in D to our newly introduced unadjust operation. 
+3. We have also made changes to the constraint generation system to take advantage of the two new operations. In the original implementation, the constraint generation system will always adjust the bitwidth of the result of an operation. We have made changes to the system to not adjust `add_unadjust` and `mul_unadjust` operations. We also proved that such change will not impact the correctness of the system.
+
+## What were the hardest parts to get right
+The hardest part is to make sure the compiler works properly using our newly introduced operation. We encounter many bugs while implementing the optimization. For instance, before since the author always adjust the bitwidth after all types of operations, the bitwidth is fixed. However, now since an operation might be using unadjusted bitwidth the logic of calculating the bitwidth of the current element is more complicated and requires more steps to ensure an operation does not overflow.
```

Comment:
Can you please include at least one example of such a bug? This sounds really interesting, and having some concrete evidence showing exactly what can go wrong would help people understand what was interesting about this project.